example/bench/parameters.cpp

```diff
@@ -0,0 +1,111 @@
+}
+unsigned bench_params::expected_events() const {
+    return expected_spikes() * network.fan_in;
+}
+unsigned bench_params::expected_events_per_interval() const {
+    return expected_spikes_per_interval() * network.fan_in;
+}
+
+std::ostream& operator<<(std::ostream& o, const bench_params& p) {
+    o << "benchmark parameters:\n"
+      << "  name:          " << p.name << "\n"
+      << "  num cells:     " << p.num_cells << "\n"
+      << "  duration:      " << p.duration << " ms\n"
+      << "  fan in:        " << p.network.fan_in << " connections/cell\n"
+      << "  min delay:     " << p.network.min_delay << " ms\n"
+      << "  spike freq:    " << p.cell.spike_freq_hz << " Hz\n"
+      << "  cell overhead: " << p.cell.realtime_ratio << " ms to advance 1 ms\n";
+    o << "expected:\n"
+      << "  cell advance: " << p.expected_advance_time() << " s\n"
+      << "  spikes:       " << p.expected_spikes() << "\n"
+      << "  events:       " << p.expected_events() << "\n"
+      << "  spikes:       " << p.expected_spikes_per_interval() << " per interval\n"
+      << "  events:       " << p.expected_events_per_interval()/p.num_cells << " per cell per interval";
+    return o;
+}
+
+using arb::util::optional;
+using nlohmann::json;
+
+// Search a json object for an entry with a given name.
+// If found, return the value and remove from json object.
+template<typename T>
+optional<T> extract(const char* name, json& j) {
+    auto it = j.find(name);
+    if (it==j.end()) {
+        return arb::util::nullopt;
+    }
+    T value = std::move(*it);
+    j.erase(name);
+    return std::move(value);
+}
+
+bench_params read_options(int argc, char** argv) {
+    bench_params params;
+    if (argc<2) {
+        std::cout << "Using default parameters.\n";
+        return params;
+    }
+    if (argc>2) {
+        throw std::runtime_error("More than command line one option not permitted.");
+    }
+
+    std::string fname = argv[1];
+    std::cout << "Loading parameters from file: " << fname << "\n";
+    std::ifstream f(fname);
+
+    if (!f.good()) {
+        throw std::runtime_error("Unable to open input parameter file: "+fname);
+    }
+
+    nlohmann::json json;
+    json << f;
+
+    if (auto o  = extract<std::string>("name", json)) {
+        params.name = *o;
+    }
+    if (auto o  = extract<unsigned>("num-cells", json)) {
+        params.num_cells = *o;
+    }
+    if (auto o  = extract<double>("duration", json)) {
+        params.duration = *o;
+    }
+    if (auto o  = extract<double>("min-delay", json)) {
+        params.network.min_delay = *o;
+    }
+    if (auto o  = extract<unsigned>("fan-in", json)) {
+        params.network.fan_in = *o;
+    }
+    if (auto o  = extract<double>("realtime-ratio", json)) {
+        params.cell.realtime_ratio = *o;
+    }
+    if (auto o  = extract<double>("spike-frequency", json)) {
+        params.cell.spike_freq_hz = *o;
+    }
+
+    for (auto it=json.begin(); it!=json.end(); ++it) {
```

Comment:
Couldn't this just be a range expression
```
for (auto it: json) {}
```

Reply:
Maybe the problem is what a `json::iterator` dereferences to?

Reply:
You are right: `json::iterator` dereferences to the `value`, not the `key`, unlike STL containers which give a `std::pair<key,value>`.

Reply:
And I want to print out all the unused keys, which would be a dead giveaway that there is a typo in the input.